### PR TITLE
Propagate fileSystem to Get*OfFileAbove intrinsic methods

### PR DESCRIPTION
**Pull request by cdmihai**
GetDirectoryNameOfFileAbove is called twice per each project evaluation to search for Directory.Build.{Props|Targets}

### CODE REVIEW

**Comment by cdmihai** [comment]
Well, improvements are quite marginal for this one.

Project | src\msbuild.master 1st run | src\msbuild.master 2nd run | Improvement | src\msbuild.mihai 1st run | src\msbuild.mihai 2nd run | Improvement
-- | -- | -- | -- | -- | -- | --
DotnetConsoleProject | 34.401 | 29.7608 | 13.48856 | 34.6137 | 29.9222 | 13.55388 |  
DotnetWebProject | 48.6614 | 37.363 | 23.2184 | 48.9981 | 37.4618 | 23.54438 |  
DotnetMvcProject | 55.222 | 39.7424 | 28.03158 | 55.2012 | 39.7505 | 27.98979 |  
Picasso | 258.1242 | 226.3237 | 12.31984 | 256.54 | 223.2164 | 12.98963 |  
SmallP2POldCsproj | 46.4664 | 42.4101 | 8.729534 | 46.622 | 42.4295 | 8.992536 |  
SmallP2PNewCsproj | 166.6562 | 135.4448 | 18.72802 | 174.0983 | 132.8031 | 23.71947 |  
LargeP2POldCsproj | 752.8776 | 700.571 | 6.947557 | 751.2062 | 695.5654 | 7.406861 |  
Generated_100_100_v150 | 1209.729 | 1197.162 | 1.038844 | 1192.789 | 1167.486 | 2.121331 |  
OrchardCore | 2264.233 | 1654.686 | 26.92068 | 2245.285 | 1631.354 | 27.34313 |  
Roslyn | 3429.948 | 2869.029 | 16.35358 | 3420.434 | 2846.373 | 16.78327 |  
WebLargeCore | 2485.726 | 1929.288 | 22.38535 | 2452.99 | 1877.997 | 23.44051 |  

### DesignTimeBuild Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 36.7149 -> 36.6435 (-0.194%)
DotnetWebProject | :red_circle: | yes | 228.024 -> 230.341 (1.016%)
DotnetMvcProject | :white_check_mark: | yes | 234.8998 -> 231.1865 (-1.581%)
Picasso | :red_circle: | yes | 1580.0867 -> 1594.6862 (0.924%)
SmallP2POldCsproj | :white_check_mark: | yes | 69.6224 -> 68.906 (-1.029%)
SmallP2PNewCsproj | :white_check_mark: | yes | 667.7962 -> 662.9372 (-0.728%)
LargeP2POldCsproj | :white_check_mark: | yes | 11873.1952 -> 11656.5876 (-1.824%)
OrchardCore | ::ok_hand: | no | 49870.8521 -> 49824.5871 (-0.093%)

### DesignTimeBuild Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 6280990 -> 6281925 (0.015%)
DotnetWebProject | :white_check_mark: | yes | 21288329 -> 18432103 (-13.417%)
DotnetMvcProject | :red_circle: | yes | 19855062 -> 22225215 (11.937%)
Picasso | :red_circle: | yes | 182404154 -> 190770650 (4.587%)
SmallP2POldCsproj | :red_circle: | yes | 8609422 -> 8624300 (0.173%)
SmallP2PNewCsproj | ::ok_hand: | no | 91924977 -> 90977263 (-1.031%)
LargeP2POldCsproj | :ok_hand: | no | 1178803330 -> 1179750990 (0.08%)
OrchardCore | :red_circle: | yes | 2649344735 -> 2718773904 (2.621%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 35.5723 -> 35.4818 (-0.254%)
DotnetWebProject | ::ok_hand: | no | 49.8711 -> 49.7872 (-0.168%)
DotnetMvcProject | ::ok_hand: | no | 56.5646 -> 56.4538 (-0.196%)
Picasso | :ok_hand: | no | 309.9398 -> 310.8799 (0.303%)
SmallP2POldCsproj | :white_check_mark: | yes | 52.9788 -> 52.5433 (-0.822%)
SmallP2PNewCsproj | :ok_hand: | no | 214.3202 -> 214.9843 (0.31%)
LargeP2POldCsproj | :white_check_mark: | yes | 936.4889 -> 927.3946 (-0.971%)
OrchardCore | :red_circle: | yes | 3332.7359 -> 3357.6386 (0.747%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 5725914 -> 5726013 (0.002%)
DotnetWebProject | :ok_hand: | no | 7740970 -> 7743134 (0.028%)
DotnetMvcProject | :ok_hand: | no | 8449754 -> 8449926 (0.002%)
Picasso | :red_circle: | yes | 10035967 -> 11411881 (13.71%)
SmallP2POldCsproj | :red_circle: | yes | 7056500 -> 7063888 (0.105%)
SmallP2PNewCsproj | ::ok_hand: | no | 15053136 -> 14795673 (-1.71%)
LargeP2POldCsproj | :ok_hand: | no | 39511750 -> 40205092 (1.755%)
OrchardCore | :ok_hand: | no | 70989782 -> 71626202 (0.896%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 29.8801 -> 30.0111 (0.438%)
DotnetWebProject | :red_circle: | yes | 37.4633 -> 37.6057 (0.38%)
DotnetMvcProject | ::ok_hand: | no | 39.8958 -> 39.8448 (-0.128%)
Picasso | :white_check_mark: | yes | 226.7592 -> 223.3359 (-1.51%)
SmallP2POldCsproj | :ok_hand: | no | 42.4807 -> 42.488 (0.017%)
SmallP2PNewCsproj | :white_check_mark: | yes | 135.2763 -> 132.9333 (-1.732%)
LargeP2POldCsproj | :white_check_mark: | yes | 701.4799 -> 696.1635 (-0.758%)
Generated_100_100_v150 | :white_check_mark: | yes | 1197.4505 -> 1167.8513 (-2.472%)
OrchardCore | :white_check_mark: | yes | 1649.4145 -> 1633.3662 (-0.973%)
Roslyn | :white_check_mark: | yes | 2862.7464 -> 2843.2493 (-0.681%)
WebLargeCore | :white_check_mark: | yes | 1936.5894 -> 1882.983 (-2.768%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 5567645 -> 5567756 (0.002%)
DotnetWebProject | :white_check_mark: | yes | 6918229 -> 6917155 (-0.016%)
DotnetMvcProject | :white_check_mark: | yes | 7377763 -> 7376620 (-0.015%)
Picasso | :white_check_mark: | yes | 36646671 -> 36630819 (-0.043%)
SmallP2POldCsproj | :white_check_mark: | yes | 6984658 -> 6976312 (-0.119%)
SmallP2PNewCsproj | :white_check_mark: | yes | 23279547 -> 23269250 (-0.044%)
LargeP2POldCsproj | ::ok_hand: | no | 37696488 -> 36918931 (-2.063%)
Generated_100_100_v150 | :red_circle: | yes | 46510125 -> 46534649 (0.053%)
OrchardCore | :white_check_mark: | yes | 61072955 -> 60927423 (-0.238%)
Roslyn | :ok_hand: | no | 92033491 -> 94774830 (2.979%)
WebLargeCore | :ok_hand: | no | 80647116 -> 80655779 (0.011%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 34.508 -> 34.7275 (0.636%)
DotnetWebProject | :red_circle: | yes | 48.7895 -> 49.0791 (0.594%)
DotnetMvcProject | ::ok_hand: | no | 55.3327 -> 55.3157 (-0.031%)
Picasso | :white_check_mark: | yes | 257.8464 -> 256.0725 (-0.688%)
SmallP2POldCsproj | :red_circle: | yes | 46.5192 -> 46.7229 (0.438%)
SmallP2PNewCsproj | ::ok_hand: | no | 170.6644 -> 170.3756 (-0.169%)
LargeP2POldCsproj | :white_check_mark: | yes | 753.5642 -> 751.3674 (-0.292%)
Generated_100_100_v150 | :white_check_mark: | yes | 1210.5778 -> 1191.783 (-1.553%)
OrchardCore | :white_check_mark: | yes | 2261.6568 -> 2246.9862 (-0.649%)
Roslyn | ::ok_hand: | no | 3425.4661 -> 3421.513 (-0.115%)
WebLargeCore | :white_check_mark: | yes | 2485.941 -> 2454.3823 (-1.269%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 5788869 -> 5788892 (0%)
DotnetWebProject | :white_check_mark: | yes | 7792765 -> 7792484 (-0.004%)
DotnetMvcProject | ::ok_hand: | no | 8498127 -> 8496812 (-0.015%)
Picasso | :ok_hand: | no | 12221157 -> 13074989 (6.987%)
SmallP2POldCsproj | :white_check_mark: | yes | 7058620 -> 7051017 (-0.108%)
SmallP2PNewCsproj | ::ok_hand: | no | 19026072 -> 18491971 (-2.807%)
LargeP2POldCsproj | ::ok_hand: | no | 38109463 -> 37941004 (-0.442%)
Generated_100_100_v150 | :ok_hand: | no | 46830815 -> 46883421 (0.112%)
OrchardCore | :ok_hand: | no | 60780638 -> 62736758 (3.218%)
Roslyn | :ok_hand: | no | 90040788 -> 91268295 (1.363%)
WebLargeCore | :ok_hand: | no | 65645185 -> 68951338 (5.036%)

File: src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs [hunk 1]
```diff
@@ -430,10 +430,118 @@ public void ContextCachesExistenceChecksInConditions(EvaluationContext.SharingPo
...
+                    switch (policy)
+                    {
+                        case EvaluationContext.SharingPolicy.Shared:
+                            searchedPath.EvaluatedValue.ShouldBe(subdirectoryFile.Path);
+                            break;
+                        case EvaluationContext.SharingPolicy.Isolated:
+                            searchedPath.EvaluatedValue.ShouldBe(
+                                evaluationCount == 1
+                                    ? subdirectoryFile.Path
+                                    : rootFile.Path);
```
Comment (rainersigwald) [inline src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs]: You've done a few tests like this and it's nice and compact but I'm worried that when a failure comes around it'll take a long time to figure out what the correct behavior was supposed to be. Should we instead have two tests with clearer execution paths?

**Comment by cdmihai** [comment]
@dotnet-bot build OSX10.13 Build for CoreCLR

**Comment by cdmihai** [comment]
@dotnet-bot build OSX10.13 Build for CoreCLR

**Comment by cdmihai** [comment]
@dotnet-bot build OSX10.13 Build for CoreCLR

**Comment by cdmihai** [comment]
@dotnet-bot test OSX10.13 Build for CoreCLR

**Pull request merged**